### PR TITLE
Enhancement backward compatibility pose

### PR DESCRIPTION
**Pull request by BenjaminDeleze**
The introduction of the `Pose` node breaks the backward compatibility.

This PR adds:

- [x] A mechanism in Webots that convert `Transform` without scale (or with the default one) into `Pose` when parsing a world. It works both for worlds and protos. If the user saves the world, the conversions will be saved for the world, but not for the protos.
- [x] A script to convert the `Transform` withtout scale from proto/worlds to `Pose`.
  - The script has several known limitations: 
    - it does not support protos with templating
    - it requires the protos/worlds to be well formated. 
    - it does not convert the parameters of the protos

I could not see a performance impact.

### CODE REVIEW

**Comment by stefaniapedrazzi** [comment]
I found that the `scale` field has not been removed from many nodes (mainly the devices) that now inherit from Pose instead of Transform (Accelerometer, Speaker, TouchSensor, etc.) `.wrl`.
I can fix it in a separate PR, but just wanted to mention it because it is somehow related to backward compatibility that will be broken.

File: scripts/converter/convert_transform_to_pose.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+"""Convert Transforms with no scale or the default one into Poses"""
+
+# usage: python convert_transform_to_pose.py /path/to/your/webots/project
+from webots_parser import WebotsParser
+from pathlib import Path
+
+import sys
+
+
+def convert_to_enu(filename):
```
Comment (omichel) [inline scripts/converter/convert_transform_to_pose.py:26]: This function name is a bit misleading, isn't it?
```suggestion
def convert_to_pose(filename):
```

File: src/webots/vrml/WbVrmlNodeUtilities.cpp [hunk 2]
```diff
@@ -481,3 +482,40 @@ QString WbVrmlNodeUtilities::exportNodeToString(WbNode *node) {
   node->write(writer);
   return nodeString;
 }
+
+// Return true if we can convert the Transform to a Pose.
+bool WbVrmlNodeUtilities::transformBackwardCompatibility(WbTokenizer *tokenizer) {
+  if (tokenizer) {
```
Comment (omichel) [inline src/webots/vrml/WbVrmlNodeUtilities.cpp:488]:
```suggestion
  if (!tokenizer)
    return true;
```

File: src/webots/vrml/WbVrmlNodeUtilities.cpp [hunk 3]
```diff
@@ -481,3 +482,40 @@ QString WbVrmlNodeUtilities::exportNodeToString(WbNode *node) {
...
+        else if (token == "]") {
+          bracketCount--;
+          if (bracketCount == 0)
+            inChildren = false;
+        }
+      } else if (token == "children") {
+        inChildren = true;
+      } else if (token == "scale") {
+        for (int i = 0; i < 3; i++) {
+          if (tokenizer->nextWord() != '1') {
```
Comment (omichel) [inline src/webots/vrml/WbVrmlNodeUtilities.cpp:506]: 1 may also be written "1.0". Shall we transform the token into a float and compare it to 1?

**Review by omichel — CHANGES_REQUESTED** [review]

File: scripts/converter/convert_transform_to_pose.py (hunk 1 above)
Comment (BenjaminDeleze) [reply scripts/converter/convert_transform_to_pose.py:26]: Ooops a bad copy/paste

File: src/webots/vrml/WbVrmlNodeUtilities.cpp [hunk 4]
```diff
@@ -481,3 +482,41 @@ QString WbVrmlNodeUtilities::exportNodeToString(WbNode *node) {
...
+      else if (token == "]") {
+        bracketCount--;
+        if (bracketCount == 0)
+          inChildren = false;
+      }
+    } else if (token == "children") {
+      inChildren = true;
+    } else if (token == "scale") {
+      for (int i = 0; i < 3; i++) {
+        if (tokenizer->nextWord().toFloat() != '1') {
```
Comment (omichel) [inline src/webots/vrml/WbVrmlNodeUtilities.cpp:508]:
```suggestion
        if (tokenizer->nextWord().toFloat() != 1) {
```

**Comment by BenjaminDeleze** [comment]
The crash I was investigating was finally not due to this PR

File: src/webots/vrml/WbVrmlNodeUtilities.cpp [hunk 5]
```diff
@@ -481,3 +482,41 @@ QString WbVrmlNodeUtilities::exportNodeToString(WbNode *node) {
...
+      else if (token == "]") {
+        bracketCount--;
+        if (bracketCount == 0)
+          inChildren = false;
+      }
+    } else if (token == "children") {
+      inChildren = true;
+    } else if (token == "scale") {
+      for (int i = 0; i < 3; i++) {
+        if (tokenizer->nextWord().toFloat() != 1) {
```
Comment (omichel) [inline src/webots/vrml/WbVrmlNodeUtilities.cpp:508]:
```suggestion
        if (tokenizer->nextWord().toFloat() != 1.0f) {
```

File: src/webots/vrml/WbNodeReader.cpp [hunk 6]
```diff
@@ -49,10 +50,14 @@ WbNode *WbNodeReader::createNode(const QString &modelName, WbTokenizer *tokenize
...
-  if (model)
-    return new WbNode(modelName, worldPath, tokenizer);
-
+  if (modelName == "Transform")
+    return WbVrmlNodeUtilities::transformBackwardCompatibility(tokenizer) ? new WbNode("Pose", worldPath, tokenizer) :
+                                                                            new WbNode("Transform", worldPath, tokenizer);
+  else {
+    WbNodeModel *const model = WbNodeModel::findModel(modelName);
+    if (model)
+      return new WbNode(modelName, worldPath, tokenizer);
```
Comment (omichel) [inline src/webots/vrml/WbNodeReader.cpp:59]:
```suggestion
    if (WbNodeModel::findModel(modelName))
      return new WbNode(modelName, worldPath, tokenizer);
```

**Pull request merged**